### PR TITLE
(draft) autocomplete MVP features under experimental

### PR DESCRIPTION
**Pull request by caroqliu**
This is a draft PR of an autocomplete mvp. [Demo](https://amp-autocomplete.firebaseapp.com/) using `toggleExperiment('amp-autocomplete', true)`.

- Override baseElement callbacks
- Read from child `script` tag
- Create basic filtering function and render filtered data as results
- Show/hide results on focus/blur and esc (sometimes)
- Support "active" item styling on mouse hover or keyboard navigate
- Add keyboard handlers (up, down, enter, esc--note esc has unexpected default behavior when input type="search")
- Support min-characters and max-entries attributes
- Create autocomplete.amp.html example document

### CODE REVIEW

File: extensions/amp-autocomplete/0.1/amp-autocomplete.js [hunk 1]
```diff
@@ -14,51 +14,374 @@
...
+    if (this.maxEntries_ && this.maxEntries_ < filteredData.length) {
+      filteredData = filteredData.slice(0, this.maxEntries_);
+    }
+
+    return filteredData;
+  }
+
+  /** Set container_ visibility to visible. */
+  showResults() {
+    setStyle(this.container_, 'visibility', 'visible');
```
Comment (cvializ) [inline extensions/amp-autocomplete/0.1/amp-autocomplete.js]: I think the `visibility` property might not be the best mechanism to use to hide the results list. I would first recommend using `src/style.js#toggle`

https://github.com/ampproject/amphtml/blob/e194f8e0726b8a3241dcac2001c43669a3df88ae/src/style.js#L214-L219

File: extensions/amp-autocomplete/0.1/amp-autocomplete.js [hunk 2]
```diff
@@ -14,51 +14,374 @@
...
+        `${TAG} requires "filter" attribute.`);
+    userAssert(isEnumValue(FilterType, this.filter_),
+        `Unexpected filter: ${this.filter}`);
+
+    this.minChars_ = this.element.hasAttribute('min-characters') ?
+      parseInt(this.element.getAttribute('min-characters'), 10) : 1;
+    this.maxEntries_ = parseInt(this.element.getAttribute('max-entries'), 10);
+
+    this.container_ = this.createContainer_();
     this.element.appendChild(this.container_);
```
Comment (cvializ) [inline extensions/amp-autocomplete/0.1/amp-autocomplete.js]: We should probably meet to talk about the `this.measureMutateElement` function sometime. It's a very important method for writing performant DOM reads/writes.

File: extensions/amp-autocomplete/0.1/amp-autocomplete.js [hunk 3]
```diff
@@ -14,51 +14,374 @@
...
+   * @param {string} input
+   * @return {!Array<string>}
+   * @private
+   */
+  filterData_(data, input) {
+    let filteredData = data.filter(item => {
+      switch (this.filter_) {
+        case FilterType.SUBSTRING:
+          return item.includes(input);
+          break;
```
Comment (alanorozco) [inline extensions/amp-autocomplete/0.1/amp-autocomplete.js]: (Sorry, drive-by.) `break` is not needed after return or throwing. Saves you a few bytes :)

Comment (cvializ) [reply extensions/amp-autocomplete/0.1/amp-autocomplete.js]: 🚗

File: extensions/amp-autocomplete/0.1/amp-autocomplete.js [hunk 4]
```diff
@@ -14,51 +14,374 @@
...
+   * Create and return <div> element from given plan-text item.
+   * @param {string} item
+   * @return {!HTMLElement}
+   * @private
+   */
+  createElementFromItem_(item) {
+    const element = this.element.ownerDocument.createElement('div');
+    element.classList.add('i-amphtml-autocomplete-item');
+    element.setAttribute('role', 'listitem');
+    element.addEventListener('mousedown', this.mousedownHandler_.bind(this));
```
[the remark this answers is not in the document]
Comment (caroqliu) [reply extensions/amp-autocomplete/0.1/amp-autocomplete.js]: This is such a great optimization suggestion! It is horrible that it was going to create an event listener for every item each time the user changes their input. :astonished:

[the remark this answers is not in the document]
Comment (caroqliu) [reply extensions/amp-autocomplete/0.1/amp-autocomplete.js]: Is `isItemElement` a future-proofing check? Right now I don't think there is anywhere inside the container that the user could click and have the target not be an item.

File: extensions/amp-autocomplete/0.1/amp-autocomplete.js [hunk 5]
```diff
@@ -14,51 +14,374 @@
...
+        }
+        break;
+      case Keys.UP_ARROW:
+        if (this.resultsShowing()) {
+          this.updateActiveItem_(-1);
+        }
+        break;
+      case Keys.ENTER:
+        if (this.activeElement_) {
+          event.preventDefault();
```
[the remark this answers is not in the document]
Comment (caroqliu) [reply extensions/amp-autocomplete/0.1/amp-autocomplete.js]: Yes, that's true. I'll add a note for that.

File: extensions/amp-autocomplete/0.1/amp-autocomplete.css [hunk 6]
```diff
@@ -13,3 +13,32 @@
...
+    flex-direction: column;
+    align-items: stretch;
+ }
+
+.i-amphtml-autocomplete-input {
+    padding: 4px;
+ }
+
+ .i-amphtml-autocomplete-results {
+    margin-top: 4px;
```
[the remark this answers is not in the document]
Comment (andrewwatterson) [reply extensions/amp-autocomplete/0.1/amp-autocomplete.css]: Huh yeah I also haven't seen this before. I can't think of any reason it would be bad, but I'd also have erred on the side of `position: absolute`. Two additional things:

a) Carlos' concern about 4px of jumping can be solved by taking the margin-top off `-autocomplete-results` (which is the thing with height: 0) and putting it on `.i-amphtml-autocomplete-item:first-child`

b) One thing to consider here is where the results div is parented. Parenting it here is *much* cleaner (and may be necessary for a11y), but has the disadvantage that it can be cut off by any ancestors with `overflow: hidden`. I wouldn't *expect* AMP pages to have complicated layouts that would make this a problem, but it's also nearly impossible to work around if you have a case that needs `overflow: hidden`.
